pkg/controller/clusterstorage/clusterstorage_controller.go

```diff
@@ -183,7 +184,34 @@ func (r *ReconcileClusterStorage) Reconcile(request reconcile.Request) (reconcil
 		return reconcile.Result{}, err
 	}
 
-	// StorageClass already exists - don't requeue
+	// Check if this StorageClass needs to be updated. As many of the storage class attributes require
+	// recreation, examine the AllowVolumeExpansion and MountOptions for differences.
+	needUpdate := false
+	if sc.AllowVolumeExpansion != found.AllowVolumeExpansion {
+		found.AllowVolumeExpansion = sc.AllowVolumeExpansion
+		needUpdate = true
+	}
+
+	if !reflect.DeepEqual(sc.MountOptions, found.MountOptions) {
+		found.MountOptions = sc.MountOptions
+		needUpdate = true
+	}
```

Comment:
If I understand correctly, if the user has `mountOptions` set, we overwrite it with an empty value (since `mountOptions` is not set in our yaml file for 4.2). The same happens for `allowVolumeExpansion`.

However, if the user has deliberately chosen to set `volumeBindingMode` to `Immediate` in 4.1, we won't make it `WaitForFirstConsumer`, which is also the default for 4.2.  In other words, we enforce updating **some** fields, but not **others**.

I think ideally we should make the operator the "owner" of this SC, always updating it to the version we ship in the yaml file.

Reply:
We could achieve that by:

1) deleting and re-creating the storageClass when they differ. I'm not sure if this is acceptable (a user might create a PVC during the re-creation process).

2) Always update the storageClass, but making sure we don't overwrite fields that cannot be changed (e.g., `parameters`). This is only to cover cases where the user deleted the SC in 4.1 and re-created it changing only a few fields (like `parameters`).

@jsafrane @gnufied: thoughts?

Reply:
I agree with @bertinatto, the operator should probably own the storage class it crates and try to overwrite user changes in the fields that can be changed.

We should document that the default storage class is owned by OpenShift and users should create their own one if they don't like ours.

Reply:
To change `volumeBindingMode` we'd have to recreate the class, correct? Because the field is immutable from https://github.com/kubernetes/kubernetes/blob/8126201b73b3fa52bd376fc98cb00637af493b72/pkg/apis/storage/validation/validation.go#L75 ?

I've moved this to an internal function and updated the PR to update the 3 fields that I currently see we can change.

I definitely agree that we'll need a documentation PR as well.

Reply:
if we delete and re-create the SC then we will lose annotation (set by user) that disables this SC to be default SC and it should be possible to disable this SC as default SC. 

If we are going to delete and re-create the SC then at very minimum we should retain previous value of `storageclass.kubernetes.io/is-default-class` annotation.  But can a case be made that why should `storageclass.kubernetes.io/is-default-class` be treated as special?

I guess this is why all objects managed by an operator should have `UnsupportedConfigOverrides` so as users don't directly modify objects created by the operator but by not having a CR or api for this, makes it harder. 


For now - my vote is delete and re-create the SC but keep default SC annotation to previous value.

Reply:
I've updated the PR to check for differences between the configurable fields, and if one is found to delete and recreate the StorageClass. This should now keep the annotations from the previous StorageClass.